http/src/handler.rs

```diff
@@ -70,87 +63,227 @@ impl<M: Metadata, S: Middleware<M>> server::Handler<HttpStream> for ServerHandle
-	metadata: Option<M>,
-	control: Control,
-	request: Request,
-	response: Response,
-	continue_on_invalid_cors: bool,
-	/// Asynchronous response waiting to be moved into `response` field.
-	waiting_sender: mpsc::Sender<Response>,
-	waiting_response: mpsc::Receiver<Response>,
+	state: RpcHandlerState<M>,
+	is_options: bool,
+	cors_header: cors::CorsHeader<header::AccessControlAllowOrigin>,
+}
+
+impl<M: Metadata, S: Middleware<M>> Future for RpcHandler<M, S> {
+	type Item = server::Response;
+	type Error = hyper::Error;
+
+	fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+		let new_state = match mem::replace(&mut self.state, RpcHandlerState::Done) {
+			RpcHandlerState::ReadingHeaders { request, cors_domains, continue_on_invalid_cors, } => {
+				// Read cors header
+				self.cors_header = utils::cors_header(&request, &cors_domains);
+				self.is_options = *request.method() == Method::Options;
+				// Read other headers
+				RpcPollState::Ready(self.read_headers(request, continue_on_invalid_cors))
+			},
+			RpcHandlerState::ReadingBody { body, request, metadata, } => {
+				self.read_body(body, request, metadata)?
+			},
+			RpcHandlerState::Waiting(mut waiting) => {
+				match waiting.poll() {
+					Ok(Async::Ready(response)) => {
+						RpcPollState::Ready(RpcHandlerState::Writing(match response {
+							// Notification, just return empty response.
+							None => Response::ok(String::new()),
+							// Add new line to have nice output when using CLI clients (curl)
+							Some(result) => Response::ok(format!("{}\n", result)),
+						}.into()))
+					},
+					Ok(Async::NotReady) => RpcPollState::NotReady(RpcHandlerState::Waiting(waiting)),
+					Err(_) => RpcPollState::Ready(RpcHandlerState::Writing(Response::internal_error())),
+				}
+			},
+			state => RpcPollState::NotReady(state),
+		};
+
+		let (new_state, is_ready) = new_state.decompose();
+		match new_state {
+			RpcHandlerState::Writing(res) => {
+				let mut response: server::Response = res.into();
+				let cors_header = mem::replace(&mut self.cors_header, cors::CorsHeader::Invalid);
+				Self::set_response_headers(response.headers_mut(), self.is_options, cors_header.into());
+				Ok(Async::Ready(response))
+			},
+			state => {
+				self.state = state;
+				if is_ready {
+					self.poll()
+				} else {
+					Ok(Async::NotReady)
+				}
+			},
+		}
+	}
 }
 
 impl<M: Metadata, S: Middleware<M>> RpcHandler<M, S> {
-	fn response_headers(&self, is_options: bool, cors_header: Option<header::AccessControlAllowOrigin>) -> Headers {
-		let mut headers = Headers::new();
-		headers.set(self.response.content_type.clone());
+	fn read_headers(
+		&self,
+		request: server::Request,
+		continue_on_invalid_cors: bool,
+	) -> RpcHandlerState<M> {
+		if self.cors_header == cors::CorsHeader::Invalid && !continue_on_invalid_cors {
+			return RpcHandlerState::Writing(Response::invalid_cors());
+		}
+		// Read metadata
+		let metadata = self.jsonrpc_handler.extractor.read_metadata(&request);
```

Comment:
Headers are parsed on demand, you would need to know in advance what headers are required by specific extractor (possible but IMHO unnecessary).